integrate with VirtualFile changes

eed3si9n opened the pull request:
Ref https://github.com/sbt/zinc/pull/712

eed3si9n committed:
integrate with VirtualFile changes

Ref https://github.com/sbt/zinc/pull/712

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1943/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1943/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@5da4fc07c754f95e6ccb0337f34ab8b745a21b92
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@621d89fb018f3f81e40903ad9ec5a9d7a3c16c9a
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1945/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1945/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@17564bfd0101211dcabcdcfd27ba9ebe1129e17c
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@621d89fb018f3f81e40903ad9ec5a9d7a3c16c9a
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1950/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1950/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@19062822a287d2253302588f1d55577d7ef53cc9
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@a9d0167ebf663e7801e4e3447d089dc2e1ce579c
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1951/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1951/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@366e77493296bcf653510575e5492c2fd1ff9b44
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@a9d0167ebf663e7801e4e3447d089dc2e1ce579c
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n committed:
Mark Dotty as pending

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1953/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1953/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@bd98dcc36ddfad044d8141408bc692e04997757a
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@e732a055ef5d8eb6fade04bc5905ef95d17f8fc6
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

typesafe-tools commented:
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1957/console">is in progress...</a>

typesafe-tools commented:
[//]: # (Refs: pull/5417/head-pull/712/head-develop-develop-develop)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/1957/">has checked</A> the following projects, tested using sources.sh.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/pull/5417">pull/5417/head      </A> | sbt/sbt@6992ee7cef3073a38fd1d1f240f3630618b52c9d
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/712">pull/712/head       </A> | sbt/zinc@d79dcf12ab52795460a8008f5b51a5415121e089
  io                | <A HREF="https://github.com/sbt/io/tree/develop">develop             </A> | sbt/io@fcedb90e5b0480fffc85d8ebf1a4c6efadd2de94
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/develop">develop             </A> | sbt/librarymanagement@3132a3eb59fcfbabf33683388377b39a5458457d
  util              | <A HREF="https://github.com/sbt/util/tree/develop">develop             </A> | sbt/util@098796cdff7305a4aaadea9fdbba7fb6c23ff3ed

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=pull/5417/head&zinc=pull/712/head&io=develop&librarymanagement=develop&util=develop&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

eed3si9n committed:
Contraband 0.4.6

eed3si9n committed:
Update source deps

eed3si9n committed:
Comment out external hooks

Zinc now uses farm hash to invalidate the virtual paths. To use watch to detect initial changes, we need to revalidate using content hash.

eed3si9n committed:
Zinc 1.4.0-M3

eed3si9n committed:
Mima

eed3si9n committed:
Mark scala-js-macro pending

This requires workaround from Zinc

eed3si9n committed:
Mark nio/external-hooks as pending

eatkins commented:
This PR introduces a significant performance regression as measured by [scala-build-watch-performance](https://github.com/eatkins/scala-build-watch-performance). I got the following results with a custom build off of origin/develop that reverted this change:
 project | min (ms) | max (ms) | median (ms) | total (ms) | cpu % |
:------- | :------: | :------: | :-------: | :--------: | :---: |
sbt-1.3.10-turbo (3 source files) | 193 | 297 | 208 | 41835 | 2.2
sbt-1.3.10-turbo (5003 source files) | 702 | 889 | 737 | 104495 | 2.6

and I got the following when I restored this PR:
 project | min (ms) | max (ms) | median (ms) | total (ms) | cpu % |
:------- | :------: | :------: | :-------: | :--------: | :---: |
sbt-1.3.0-turbo (3 source files) | 215 | 342 | 286 | 51483 | 2.7
sbt-1.3.0-turbo (5003 source files) | 996 | 2060 | 1120 | 162077 | 2.3

I reran the tests because I saw that sbt stopped using an optimized external hooks implementation and I was curious what the impact would be. I also noticed that the `nio/external-hooks` test was marked pending which seems unfortunate because it is a regression test against a real bug that I have experienced.

I ran the test with jdk14 and with 
```
java -classpath target/classes:src/main/resources:lib/swoval/file-tree-views-2.1.1.jar build.performance.Main -i 50 -w 50 -e 5000 sbt-1.3.10-turbo
```
after changing `src/main/resources/sbt-1.3.10/project/build.properties` to `1.4.0-SNAPSHOT`.

eed3si9n commented:
Thanks for this. I've been working on bringing Zinc benchmarks up to date to measure this kind of things - https://github.com/sbt/zinc/pull/767
Though it might also be good to run some benchmark on sbt/sbt side.

I needed to comment out the external hooks in general to resume incremental compilation from a build from a different path. With the hook it was invalidating all files. I am guessing that it internally knows about absolute paths, and I need to somehow change that to a virtual one.

smarter commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -173,7 +176,33 @@ object Defaults extends BuildCommon {
+        val boot = app.provider.scalaProvider.launcher.bootDirectory
+        val ih = app.provider.scalaProvider.launcher.ivyHome
+        val coursierCache = csrCacheDirectory.value
+        val javaHome = Paths.get(sys.props("java.home"))
```
Does that still work in java9+ where the jdk classes come from `jrt:/` ?

eed3si9n replied:
Sort of but not elegantly. We ignore `rt.jar`-derived invalidation, so that seems to work at least for now. JAVA_HOME was still needed because there are other JARs that needs remapping.
